pkg/orchestration/wiring/wiringplugin/shapes.go

```diff
@@ -39,3 +46,30 @@ type ShapeMeta struct {
 type BindableShapeStruct struct {
 	ServiceInstanceName ProtoReference `json:"serviceInstanceName"`
 }
+
+// CopyShape copies source shape into the target shape.
+// The main purpose of this function is to convert between typed and UnstructuredShape representations
+// of the same shape.
+func CopyShape(source, target Shape) error {
+	st := reflect.TypeOf(source)
+	tt := reflect.TypeOf(target)
+	switch {
+	case st == tt:
+		// Same type already, just deep copy
+		sv := reflect.ValueOf(source)
+		tv := reflect.ValueOf(target)
+		method := sv.MethodByName("DeepCopyInto")
+		method.Call([]reflect.Value{tv})
+	default:
+		// Convert by round-tripping to and from JSON
+		data, err := json.Marshal(source)
+		if err != nil {
+			return errors.WithStack(err)
```

Comment:
wrap these errors with message to make it more obvious if marshalling or unmarshalling failed